### PR TITLE
feat: support regex for `prerender.ignore`

### LINKED ISSUE #308: allow excluding routes from prerenderer crawling
Would be nice if something is intended to be dynamic.

Perhaps we could support `prerender.exclude` with RegExp, function or string signature? Or a render hook with option to skip rendering (https://github.com/unjs/nitro/issues/88)?

### PR DESCRIPTION
**Pull request by thezzisu**
### 🔗 Linked issue

Resolves #308

### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Enhance prerender.ignore with the ability to use RegExp.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by thezzisu** [comment]
Notice: maybe conflict with #1032

**Comment by Atinux** [comment]
Will first merge #1032 to rebase your branch

**Comment by Atinux** [comment]
Fixed the conflicts and improve the docs at the same time.

File: src/prerender.ts [hunk 1]
```diff
@@ -112,7 +112,11 @@ export async function prerender(nitro: Nitro) {
 
     // Check for explicitly ignored routes
     for (const ignore of nitro.options.prerender.ignore) {
-      if (route.startsWith(ignore)) {
+      if (typeof ignore === "string") {
+        if (route.startsWith(ignore)) {
+          return false;
+        }
+      } else if (ignore.test(route)) {
```
Comment (Atinux) [inline src/prerender.ts:119]: Shall we test if `ignore.test` is a function to make sure it does not break anything?

**Comment by Atinux** [comment]
Wondering if we could add a test case for it as well.

**Comment by pi0** [comment]
I'm thinking we might be better first allow `**` and `*` in ignore patterns to keep it consistent with route rules. We can also introduce a hook / function filter to allow ignoring routes on demand for more advanced cases.

**Comment by Atinux** [comment]
Allowing the pattern `**` and `*` will solve your usecase @thezzisu ?

**Comment by Atinux** [comment]
So what are the next steps?

File: src/prerender.ts (hunk 1 above)
Comment (thezzisu) [reply src/prerender.ts:119]: For now, ignore would be neither string or Regex if the provided configuration is checked by TypeScript. So there's no need to test wether `ignore.test` exists.

**Comment by thezzisu** [comment]
As @pi0 said, custom hooks/callbacks would be the most powerful solution; but, for most use cases, RegEx should be enough.

> Allowing the pattern `**` and `*` will solve your usecase @thezzisu ?

In fact, my use cases needs the power of "excluded ignore", which is not easily solvable with route rules matching. The underlying library, whose radix-tree-matching do not produce result with the same order as in the input.

**Comment by thezzisu** [comment]
> Wondering if we could add a test case for it as well.

Should I implement more strict input checking? If the user uses TypeScript, the current solution will work fine.

**Comment by Atinux** [comment]
I think it's fine for types using TS.

If I understand correctly, you don't want to support regex but the ability to give a function instead @pi0 ?

**Comment by thezzisu** [comment]
> A hook would be nice for advanced usecases.
> 
> I prefer to keep `ignore` strict string pattern format as much as possible. This (same as route rules) is an advantage to be able static rules to provider rules which by introducing regexes to route defs we lose.
> 
> What do you think about introducing `prerender.include` as well:
> 
> ```js
> prerender: {
>   include: ['/static/**'],
>   exclude: ['/**']
> }
> ```

But how about the priority? Is the 'include' a 'negative exclude' like `!blahblah` syntax in `.gitignore`?

Since this option won't affect how route options work (in fact, it's a option for the crawler, not the server), I think RegExs should be fine.

**Comment by thezzisu** [comment]
> > But how about the priority?
> 
> The most precise rule takes into priority
> 
> > Since this option won't affect how route options work
> 
> The point is route rule definition consistency and having most simple way to do things when possible.

OK. How about add a `prerenderRules` like `routeRules` that using the similar syntax `Record<string, SomeOptions>`?

**Comment by thezzisu** [comment]
> > > But how about the priority?
> > 
> > 
> > The most precise rule takes into priority
> > > Since this option won't affect how route options work
> > 
> > 
> > The point is route rule definition consistency and having most simple way to do things when possible.
> 
> OK. How about add a `prerenderRules` like `routeRules` that using the similar syntax `Record<string, SomeOptions>`?

The reason is that, routeRules is **for server**, but the prerender is a standalone thing though.

**Comment by pi0** [comment]
We do actually support `prerender: true` for `routeRules` (route rules are unified routing rules not specific to server).

We can indeed support `prerender: false` to allow overiridng for specific patterns 👍🏼

**Comment by thezzisu** [comment]
> We do actually support `prerender: true` for `routeRules` (route rules are unified routing rules not limited to server).
> 
> We can indeed support `prerender: false` to allow overiridng for specific patterns 👍🏼

So the problem is I want to make `/` not prerendered (SSR) when using the server built by nuxt build, but to be prerendered when using nuxt generate.

**Comment by pi0** [comment]
Yes that would be possible! (SSR renderer won't be affected by `prerender` rule only `nitro prerender`/`nuxt generate` uses this specific rule).

**Comment by thezzisu** [comment]
Sorry for the delay. Now the problems becomes that the https://github.com/unjs/radix3#route-matcher is not matching the most specific rule! Let's see a example below:

image
image

Simple swapping the rules will affect the output.

**Comment by Atinux** [comment]
Nice catch @thezzisu 

It seems we need a way to prioritize the routes for best matching strategy, similar to what vue-router 4 is doing 🤔

**Comment by thezzisu** [comment]
I think this would be a better solution than supporting regex. I'll close this pull request, and open a issue on [unjs/radix3](https://github.com/unjs/radix3#route-matcher) instead.

**Comment by pi0** [comment]
Radix 3 basically does that :) but we don't use it internally in nitro prerender inference...

**Comment by Atinux** [comment]
So open back an issue in Nitro pre-render to support this I guess 😅

**Comment by pi0** [comment]
~> https://github.com/unjs/nitro/issues/1078

**Pull request closed without merging**